### PR TITLE
InvalidLinkMode: changed invalid link handling

### PR DESCRIPTION
**Pull request by klimesf**
[See the RFC](http://forum.nette.org/en/22153-link-and-plink-macros-produce-e-warning-if-link-is-not-valid)

The build isn't failing due to the changes I made. I can't address the issue in this PR.

### CODE REVIEW

File: tests/Application.DI/ApplicationExtension.scan.phpt [hunk 1]
```diff
@@ -89,3 +89,139 @@ test(function() {
...
+	eval($code);
+
+	/** @var Nette\DI\Container $container */
+	$container = new Container7;
+	$tags = $container->findByTag('nette.presenter');
+	Assert::count( 1, array_keys($tags, 'Presenter1') );
+	$tmp = array_keys($tags, 'Presenter1');
+	$presenter = $container->getService($tmp[0]);
+	Assert::same( \Nette\Application\UI\Presenter::INVALID_LINK_SILENT, $presenter->invalidLinkMode );
+});
```
Comment (klimesf) [inline tests/Application.DI/ApplicationExtension.scan.phpt]: I will refactor this, WIP

File: src/Application/UI/Presenter.php [hunk 2]
```diff
@@ -1055,19 +1058,21 @@ public static function argsToParams($class, $method, & $args, $supplemental = ar
...
-		if ($this->invalidLinkMode === self::INVALID_LINK_SILENT) {
-			return '#';
-
-		} elseif ($this->invalidLinkMode === self::INVALID_LINK_WARNING) {
-			return '#error: ' . $e->getMessage();
-
-		} else { // self::INVALID_LINK_EXCEPTION
-			throw $e;
+		switch($this->invalidLinkMode) {
+			case self::INVALID_LINK_WARNING:
```
Comment (fprochazka) [inline src/Application/UI/Presenter.php]: Exception should not be dropped, it should produce at least a warning. Untill it's removed few versions later.

File: src/Application/UI/Presenter.php [hunk 3]
```diff
@@ -1055,19 +1058,21 @@ public static function argsToParams($class, $method, & $args, $supplemental = ar
...
+				trigger_error('Invalid link: ' . $e->getMessage(), E_USER_WARNING);
+			case self::INVALID_LINK_VISUAL:
+				return '#error: ' . $e->getMessage();
+			case self::INVALID_LINK_SILENT:
+			default:
+				trigger_error('Invalid link: ' . $e->getMessage(), E_USER_WARNING);
```
Comment (fprochazka) [inline src/Application/UI/Presenter.php]: why is silent producing an error?

File: src/Bridges/ApplicationDI/ApplicationExtension.php [hunk 4]
```diff
@@ -141,4 +145,22 @@ private function findPresenters()
 		return $presenters;
 	}
 
+
+	/**
+	 * Sets invalidLinkMode for presenters.
+	 * @param bool
+	 */
+	private function setSilentInvalidLinks($silent = TRUE)
```
Comment (fprochazka) [inline src/Bridges/ApplicationDI/ApplicationExtension.php]: I don't like that you've used a magic setter for this.

File: src/Application/UI/Presenter.php (hunk 3 above)
Comment (klimesf) [inline src/Application/UI/Presenter.php]: This is described in the RFC. SILENT is meant for production and the name refers to the text it returns. E_USER_WARNING is produced so it's visible in the logs.

File: src/Application/UI/Presenter.php (hunk 2 above)
Comment (klimesf) [inline src/Application/UI/Presenter.php]: It behaves exactly like the INVALID_LINK_SILENT, see the `default` keyword below.

Comment (fprochazka) [inline src/Application/UI/Presenter.php]: I don't like that it's not explicit.

File: src/Application/UI/Presenter.php (hunk 3 above)
Comment (fprochazka) [inline src/Application/UI/Presenter.php]: I see now :)

File: src/Bridges/ApplicationDI/ApplicationExtension.php [hunk 5]
```diff
@@ -141,4 +145,22 @@ private function findPresenters()
 		return $presenters;
 	}
 
+
+	/**
+	 * Sets invalidLinkMode for presenters.
+	 * @param bool
+	 */
+	private function configureInvalidLinkMode($silent = true)
```
Comment (klimesf) [inline src/Bridges/ApplicationDI/ApplicationExtension.php]: @fprochazka Is it better this way? :)

**Comment by klimesf** [comment]
If Presenter::invalidLinkMode is set to something else than the provided constants, it will now behave the same as INVALID_LINK_SILENT, whereas before it would behave as INVALID_LINK_EXCEPTION. Do you think this is a BC break?

File: src/Bridges/ApplicationDI/ApplicationExtension.php (hunk 5 above)
Comment (kubinpetr) [inline src/Bridges/ApplicationDI/ApplicationExtension.php]: Use TRUE instead of true.

Comment (klimesf) [inline src/Bridges/ApplicationDI/ApplicationExtension.php]: Thank you! I missed this one.

**Comment by dg** [comment]
What about something like this #66?

**Pull request closed without merging**